### PR TITLE
Enable multiple templates per repository

### PR DESCRIPTION
**Pull request by RomHartmann**
For issue https://github.com/cookiecutter/cookiecutter/issues/1223

- [x] Implemented cli feature to find `cookiecutter.json` in a subdirectory.  This was written to solve the problem that cookiecutter could not work with subdirectories within github.  Also works with local, though that use case makes no sense :)
- [x] Updated unit tests for new cli option
- [x] Added a unit test to read from a subdir
- [x] Added documentation on how to use.  Not sure where that should go though.
- [x] Manually tested it works with local path
- [x] Manually tested it works with local zip file
- [x] Manually tested it works with github

### CODE REVIEW

File: docs/advanced/repo_subdirectories.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Organizing cookiecutter templates in subdirectories
```
Comment (insspb) [inline docs/advanced/repo_subdirectories.rst]: Should be linked in ``docs/advanced/index.rst``

File: docs/advanced/repo_subdirectories.rst [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        ├── subdirectory-name/
+            ├── {{cookiecutter.project_slug}}/
+            └── cookiecutter.json
+
+And then, using the cli
+
+    cookiecutter https://github.com/user/repo-name.git --subdirectory="subdirectory-name"
+
+
+Cookiecutter clones the repo into `~/.cookiecutter` as usual, then `cd`s
```
Comment (insspb) [inline docs/advanced/repo_subdirectories.rst]: This does not explain what user get's at the end. As far as I can understand it is half way only.

Comment (insspb) [reply docs/advanced/repo_subdirectories.rst]: Will it work wiz zip files?

**Comment by RomHartmann** [comment]
@insspb Should I also increment version in setup.py?

File: docs/advanced/repo_subdirectories.rst (hunk 1 above)
Comment (RomHartmann) [inline docs/advanced/repo_subdirectories.rst]: @insspb I see other features have for example `Dictionary Variables (1.5+)`.  Should I include this?

**Comment by insspb** [comment]
> @insspb Should I also increment version in setup.py?

No, unfortunately, we have a long new version deploy process, so you should not increment it, as until several core developers review this it is hard to say when it will be merged. (Last PRs merged fast, but we have some dev depth with opened PRs since 2017)

File: docs/advanced/repo_subdirectories.rst (hunk 1 above)
Comment (insspb) [reply docs/advanced/repo_subdirectories.rst]: Yes, my review note was mention that you should add this document to index toc.

File: docs/advanced/repo_subdirectories.rst (hunk 2 above)
Comment (RomHartmann) [reply docs/advanced/repo_subdirectories.rst]: I'll reword it to be more explicit.  Yup, it works with local, zip, github (updated OP with more checkboxes)

**Comment by RomHartmann** [comment]
@insspb also added unit test so that coverage should now not go down anymore.

Please let me know if there is anything I can do to help speed up merging of PR.  I really want to start using this feature (and I think another feature being reviewed right now https://github.com/cookiecutter/cookiecutter/pull/1173 😄 )

File: docs/advanced/index.rst [hunk 3]
```diff
@@ -20,3 +20,4 @@ Various advanced topics regarding cookiecutter usage.
    choice_variables
    dict_variables
    template_extensions
+   subdirectories
```
Comment (insspb) [inline docs/advanced/index.rst]:
```suggestion
   repo_subdirectories
```
Do not bring new errors to docs :)

P.S. I know that docs has some not-rendered endpoints right now. Such errors already fixed at #1216

File: docs/advanced/repo_subdirectories.rst [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            ├── {{cookiecutter.project_slug}}/
+            └── cookiecutter.json
+
+And then, using the cli::
+
+    cookiecutter https://github.com/user/repo-name.git --subdirectory="subdirectory-name"
+
+Cookiecutter follows this high-level algorithm
+
+1) Copy/clone/unzip a directory into (by default) ``~/.cookiecutter/``
```
Comment (insspb) [inline docs/advanced/repo_subdirectories.rst]: Formatting of this block looks very strange after Sphinx rendering.
You can render docs locally to look how it will be on docs site.
Here is screenshot.

Screenshot from 2019-09-04 12-15-39

P.S. Writting docs is hard :)

Comment (RomHartmann) [reply docs/advanced/repo_subdirectories.rst]: Thanks, fixed!

**Comment by insspb** [comment]
@RomHartmann checked docs on local and get another errors after it. 
Then I begin to fix it and catched that subdirectory is not correct name for this feature, as `-cd` flag too.
I made a PR https://github.com/RomHartmann/cookiecutter/pull/1  with all changes that will make it better I think.

**Comment by RomHartmann** [comment]
> @RomHartmann checked docs on local and get another errors after it.
> Then I begin to fix it and catched that subdirectory is not correct name for this feature, as `-cd` flag too.
> I made a PR [RomHartmann#1](https://github.com/RomHartmann/cookiecutter/pull/1) with all changes that will make it better I think.

@insspb Approved PR.  Don't know what the merge policy is in this project, so I'll fall back to least-permission and wait for you to merge it :)

**Comment by insspb** [comment]
@RomHartmann you should merge my PR to your RomHartmann:1223-enable_repo_subdirectories repo. Then it will be available here automatically. 
Core devs with write access can push code directly to your repo, but I do not have write access yet. So I am making PR. 

Merge policy in cookiecutter:
Reviews and approve from 2-3 core at least, after one of cores marks PR as "Needs review"

File: cookiecutter/main.py [hunk 5]
```diff
@@ -25,7 +25,7 @@
 def cookiecutter(
         template, checkout=None, no_input=False, extra_context=None,
         replay=False, overwrite_if_exists=False, output_dir='.',
-        config_file=None, default_config=False, password=None):
+        config_file=None, default_config=False, password=None, dir=None):
```
Comment (michaeljoseph) [inline cookiecutter/main.py]: Let's not shadow the [`dir` built-in function](https://docs.python.org/3.7/library/functions.html#dir) please?

File: cookiecutter/repository.py [hunk 6]
```diff
@@ -112,9 +113,14 @@ def determine_repo_dir(template, abbreviations, clone_to_dir, checkout,
         cleanup = False
 
     for repo_candidate in repository_candidates:
+        if dir:
+            repo_candidate = os.path.join(repo_candidate, dir)
         if repository_has_cookiecutter_json(repo_candidate):
             return repo_candidate, cleanup
 
+    if dir:
```
Comment (michaeljoseph) [inline cookiecutter/repository.py]: Could we move this to before L115? This means we could drop the condition on L116...

File: tests/repository/test_find_directory.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
```
Comment (michaeljoseph) [inline tests/repository/test_find_directory.py]: The current convention is to name the test filename after the function it tests, so this should be more like `test_determine_repo_dir_finds_subdirectories`?

File: docs/advanced/directories.rst [hunk 8]
```diff
@@ -0,0 +1,28 @@
+.. _directories:
+
```
Comment (michaeljoseph) [inline docs/advanced/directories.rst]: This is a great start, but needs some editing / wordsmithing improvements.

File: cookiecutter/main.py (hunk 5 above)
Comment (RomHartmann) [reply cookiecutter/main.py]: Ah, I didn't catch that...  This change was made in above PR (https://github.com/cookiecutter/cookiecutter/pull/1224/commits/5a6982dc19330a111ce08075cbf84663ab797284)
Shall I change it back to `subdir`?

File: docs/advanced/directories.rst (hunk 8 above)
Comment (RomHartmann) [reply docs/advanced/directories.rst]: @michaeljoseph Is there anything in particular that you think is missing, or should I just make it generally more verbose?

File: cookiecutter/main.py (hunk 5 above)
Comment (insspb) [reply cookiecutter/main.py]: Better to make it directory. (Full name)
I saw this @michaeljoseph note, but have not time to change it yet. So if you can make it directory it will be great.

File: docs/advanced/directories.rst [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+Organizing cookiecutters in directories (1.7+)
+---------------------------------------------------
+
+*New in Cookiecutter 1.7*
+
+Cookiecutter introduces the ability to organize several templates in one
+repository or zip file, separating them by directories. This allows using
+symlinks for general files, to edit them in one place. Here is how the
```
Comment (michaeljoseph) [inline docs/advanced/directories.rst]: I think you could omit `to edit them in one place` for the same effect?

Comment (michaeljoseph) [inline docs/advanced/directories.rst]: `Here's an example repository demonstrating this feature`?

File: docs/advanced/directories.rst [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+    https://github.com/user/repo-name.git
+        ├── directory1-name/
+        |   ├── {{cookiecutter.project_slug}}/
+        |   └── cookiecutter.json
+        └── directory2-name/
+            ├── {{cookiecutter.project_slug}}/
+            └── cookiecutter.json
+
+To activate one of templates from subdirectory use cli with ``--directory`` option::
```
Comment (michaeljoseph) [inline docs/advanced/directories.rst]: `To activate one of the templates, use the ``--directory`` option` ?

File: docs/advanced/directories.rst [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+    cookiecutter https://github.com/user/repo-name.git --directory="directory-name"
```
Comment (michaeljoseph) [inline docs/advanced/directories.rst]: Let's use one of the directories from the example here?

File: docs/advanced/directories.rst [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+
+``--directory`` is a simple "transformation", where cookiecutter looks for
```
Comment (michaeljoseph) [inline docs/advanced/directories.rst]: We could probably lose this whole section?
Although it does make me wonder what would happen if you had an existing repo with `cookiecutter.json` in the root?

File: cookiecutter/main.py [hunk 13]
```diff
@@ -41,6 +41,7 @@ def cookiecutter(
     :param config_file: User configuration file path.
     :param default_config: Use default values rather than a config file.
     :param password: The password to use when extracting the repository.
+    :param directory: Subdirectory within repo where cookiecutter.json lives.
```
Comment (michaeljoseph) [inline cookiecutter/main.py]: Something more like: A relative path to a cookiecutter template in the cloned repository?

File: docs/advanced/directories.rst (hunk 8 above)
Comment (michaeljoseph) [reply docs/advanced/directories.rst]: I've made some suggestions...

**Comment by RomHartmann** [comment]
@michaeljoseph Any other change suggestions?

**Comment by SteveHNH** [comment]
Howdy folks! Interested in having this available for a project. Looks like it's all good to go but just needs to be merged?

**Comment by ryelo** [comment]
Seconding @SteveHNH, this would be great to use in our current project!

**Comment by ljvmiranda921** [comment]
So excited for this feature to be merged!
